vyper/semantics/types/user.py

```diff
@@ -281,10 +282,58 @@ def from_EventDef(cls, base_node: vy_ast.EventDef) -> "EventT":
         return cls(base_node.name, members, indexed, base_node)
 
     def _ctor_call_return(self, node: vy_ast.Call) -> None:
+        # validate keyword arguments if provided
+        if len(node.keywords) > 0:
+            return self._ctor_call_return_with_kwargs(node)
+
+        # warn about positional argument depreciation
+        msg = "Instantiating events with positional arguments is "
+        msg += "deprecated as of v0.4.1 and will be disallowed "
+        msg += "in a future release. Use kwargs instead eg. "
+        msg += "Foo(a=1, b=2)"
+
+        vyper_warn(msg, node)
+
         validate_call_args(node, len(self.arguments))
         for arg, expected in zip(node.args, self.arguments.values()):
             validate_expected_type(arg, expected)
 
+    def _ctor_call_return_with_kwargs(self, node: vy_ast.Call) -> None:
+        # TODO: Remove block when positional args are removed
+        if len(node.args) > 0:
+            # can't mix args and kwargs
+            raise InstantiationException(
+                "Event instantiation requires either all positional arguments "
+                "or all keyword arguments",
+                node,
+            )
+
+        # manually validate kwargs for better error messages instead of
+        # relying on `validate_call_args` (same as structs)
+        members = self.arguments.copy()
+        keys = list(self.arguments.keys())
+        for i, kwarg in enumerate(node.keywords):
+            # x=5 => kwarg(arg="x", value=Int(5))
+            argname = kwarg.arg
+            if argname not in members:
+                hint = get_levenshtein_error_suggestions(argname, members, 1.0)
+                raise UnknownAttribute("Unknown or duplicate event argument.", kwarg, hint=hint)
+            expected = keys[i]
+            if argname != expected:
+                raise InvalidAttribute(
+                    "Event keys are required to be in order, but got "
+                    f"`{argname}` instead of `{expected}`. (Reminder: the "
+                    f"keys in this event are {list(self.arguments)})",
+                    kwarg,
+                )
+            expected_type = members.pop(argname)
+            validate_expected_type(kwarg.value, expected_type)
```

Comment:
can we clean this code up? refactor it to be shared somehow with the struct code? if it ends up being more messy, we can declare it out of scope for this PR.

Reply:
Moved to a helper function